### PR TITLE
[ML] Inference duration and error metrics

### PR DESCRIPTION
**Pull request by prwhelan**
Add `es.inference.requests.time` metric around `infer` API.

As recommended by OTel spec, errors are determined by the presence or absence of the `error.type` attribute in the metric. "error.type" will be the http status code (as a string) if it is available, otherwise it will be the name of the exception (e.g. NullPointerException).

Additional notes:
- ApmInferenceStats is merged into InferenceStats. Originally we planned to have multiple implementations, but now we're only using APM.
- Request count is now always recorded, even when there are failures loading the endpoint configuration.
- Added a hook in streaming for cancel messages, so we can close the metrics when a user cancels the stream.

----------------------------------

Example from local node to APM (redacted a bunch):
```
{
  "_index": ".ds-metrics-apm.app.elasticsearch-default-2024.10.25-000001",
    "data_stream": {
      "dataset": "apm.app.elasticsearch",
      "namespace": "default",
      "type": "metrics"
    },
    "es": {
      "inference": {
        "requests": {
          "time": {
            "values": [
              6992.5
            ],
            "counts": [
              1
            ]
          }
        }
      }
    },
    "labels": {
      "model_id": "gpt-4o-mini",
      "otel_instrumentation_scope_name": "elasticsearch",
      "service": "openai",
      "task_type": "completion"
    },
    "numeric_labels": {
      "status_code": 200
    },
    ...
  }
}
```

### CODE REVIEW

**Comment by prwhelan** [comment]
@elasticmachine update branch

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceAction.java [hunk 1]
```diff
@@ -83,20 +90,57 @@ protected void doExecute(Task task, InferenceAction.Request request, ActionListe
...
+                taskProcessor.subscribe(instrumentedStream);
+
+                listener.onResponse(new InferenceAction.Response(inferenceResults, instrumentedStream));
+            } else {
+                listener.onResponse(new InferenceAction.Response(inferenceResults));
+                recordMetrics(model, timer, null);
+            }
+        }, e -> {
+            listener.onFailure(e);
+            recordMetrics(model, timer, e);
```
Comment (maxhniebergall) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceAction.java:129]: I think we need to do listenener.onFailure(e); after we recordMetrics. I don't think that any code after calling listener.on is guaranteed to run.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/DelegatingProcessor.java [hunk 2]
```diff
@@ -61,11 +61,14 @@ public void request(long n) {
             public void cancel() {
                 if (isClosed.compareAndSet(false, true) && upstream != null) {
                     upstream.cancel();
+                    onCancel();
                 }
             }
         };
     }
 
+    protected void onCancel() {}
```
Comment (maxhniebergall) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/DelegatingProcessor.java:70]: I'm curious why this is not an abstract method? Do we not want to implement this in all subclasses?

**Review by maxhniebergall — COMMENTED** [review]
LGTM, just a couple of questions

**Review by maxhniebergall — APPROVED** [review]
LGTM

**Comment by prwhelan** [comment]
@elasticmachine update branch

**Comment by jonathan-buttner** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/118700)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**